src/private/MDMPerformerGroup.m

```diff
@@ -76,30 +78,64 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
 #pragma mark - Private
 
 - (void)addPlan:(id<MDMPlan>)plan trace:(MDMTrace *)trace log:(MDMTransactionLog *)log {
-  [trace.committedPlans addObject:plan];
-  
+  // all named addPlan plans must first be removed before being added
+  if ([self isNamedTransactionLog:log] && !log.isRemoval) {
+    [self addPlan:plan trace:trace log:[[MDMTransactionLog alloc] initWithPlans:@[plan] target:log.target name:log.name removal:TRUE]];
+  }
+  // see if we can get a performer
+  id<MDMPerforming> performer = [self performerForPlan:plan trace:trace log:log];
+  if (performer != nil) {
+    if (plan != nil) {
+      [trace.committedPlans addObject:plan];
+    } else {
+      // this is the case whereby we are calling removePlan:named, but don't have a MDMPlan
```

Comment:
We may want two different storage units for plans in the tracer in that case, one for `committedPlans` and another for `removedPlans`.

I've found the trace notifications approach to be moderately challenging to work with, so I have a diff out that simplifies the system: http://codereview.cc/D1682. If we go with that approach instead then you can add a new method to the Tracer type for `didAddPlan(_:named:to:)`. This'll make writing the unit tests somewhat simpler.